privacyidea/api/lib/prepolicy.py

```diff
@@ -711,6 +714,30 @@ def check_max_token_user(request=None, action=None):
             if already_assigned_tokens >= max_value:
                 g.audit_object.add_policy(limit_list.get(str(max_value)))
                 raise PolicyError(ERROR)
+
+        # check maximum active tokens of user
+        limit_list = policy_object.get_action_values(ACTION.MAXACTIVETOKENUSER,
+                                                     scope=SCOPE.ENROLL,
+                                                     realm=user_object.realm,
+                                                     resolver=user_object.resolver,
+                                                     user=user_object.login,
+                                                     client=g.client_ip)
+        if limit_list:
+            # we need to check how many active tokens the user already has assigned!
+            tokenobject_list = get_tokens(user=user_object, active=True)
```

Comment:
If there are both a ``MAXTOKENUSER`` and a ``MAXACTIVETOKENUSER`` policy, we have the following issues:
* We call ``get_tokens`` two times (in lines 707 and 727), which results in database queries. We can avoid this by using:
```python
tokenobject_list = get_tokens(user=user_object)
active_tokenobject_list = [token for token in tokenobject_list if token.is_active()]
```
* We have a ``return True`` in line 707. If we hit this line, the ``MAXACTIVETOKENUSER`` policy isn't taken into account. I think this might lead to situations in which the ``MAXACTIVETOKENUSER`` policy can be bypassed?

Also, we use the "fetch number of tokens, get maximum number of allowed tokens, throw error if maximum is exceeded" pattern three times already (1x ``check_max_token_realm``, 2x ``check_max_token_user``). We could move the code to its own function:
```python
    def check_limit_exceeded(tokenobject_list, limit_list):
        already_assigned_tokens = len(tokenobject_list)
        max_value = max([int(x) for x in limit_list])
        if already_assigned_tokens >= max_value:
             g.audit_object.add_policy(limit_list.get(str(max_value)))
             raise PolicyError(ERROR)
```

This would simplify the code. What do you think?

Reply:
In regards to get_tokens it is not that simple.
The first get_tokens could not be called, since there is no MAX_TOKEN policy.
So we would have to check, if a tokenlist is already filled and then either reduce the list to an active-token-list or do a new get_tokens call. The code looks a bit cluttered. Given that most people probably will either use a max token or a max-active-token I would like to keep two get_tokens calls. (Open for discussion)

3 comments in a comment is a bit difficult to handle.

Reply:
Done the early True-exist in 3ad5a51d